riot/riot.py

```diff
@@ -62,60 +66,157 @@ def to_list(x: t.Union[_K, t.List[_K]]) -> t.List[_K]:
 _T_stdio = t.Union[None, int, t.IO[t.Any]]
 
 
-@dataclasses.dataclass(unsafe_hash=True, eq=True)
+@dataclasses.dataclass(eq=True)
 class Interpreter:
-    _T_hint = t.Union[float, int, str]
+    _path: str = dataclasses.field()
+    _version: Version = dataclasses.field(init=False)
 
-    hint: dataclasses.InitVar[_T_hint]
-    _hint: str = dataclasses.field(init=False)
+    def __post_init__(self):
+        """Get the version of the interpreter."""
+        output = subprocess.check_output(
+            [self._path, "--version"], stderr=subprocess.STDOUT
+        )
+        version = Version(output.decode().strip().split(" ")[1])
+        self._version = version
 
-    def __post_init__(self, hint: _T_hint) -> None:
-        """Normalize the data."""
-        self._hint = str(hint)
+    def __hash__(self) -> int:
+        """Return the hash of this interpreter."""
+        return hash(self._path)
+
+    def __repr__(self) -> str:
+        """Return the repr containing the path and version of this interpreter."""
+        return f"{self.__class__.__name__}('{self.version}', '{self._path}')"
 
     def __str__(self) -> str:
-        """Return the path of the interpreter executable."""
+        """Return the repr of this interpreter."""
         return repr(self)
 
+    @staticmethod
+    def _matches(v1: Version, v2: Version) -> bool:
+        """Return if v2 matches v1.
```

Comment:
It took me a few seconds to understand why that was here.

You should be a little more explicit about what you want to achieve with this one. I understand it makes sense for our use case, but it was not obvious at first read. 😉 

A "cleaner" alternative might be to have a custom `Version` class that inherits from the `packaging` `Version` and that has a new `matches` method like this one. If that's doable, it'd prevent to have this class "polluted" with this matching code altogether.

Reply:
The better approach is to have a custom `Version` class that inherits from `packaging.version.Version` and which defines a custom `Version._key`.

https://github.com/pypa/packaging/blob/47d40f640fddb7c97b01315419b6a1421d2dedbb/packaging/version.py#L292-L299

Seems like we want to compare versions but ignoring pre/post/dev/local ?

which if that is the case, then a custom key, and then relying on the built-in comparisons to `Version` should work.